sympy/polys/matrices/sdm.py

```diff
@@ -28,6 +28,37 @@ def __init__(self, elemsdict, shape, domain):
         if not all(0 <= c < n for row in self.values() for c in row):
             raise DDMBadInputError("Column out of range")
 
+    def getitem(self, i, j):
+        m, n = self.shape
+        if i >= m:
+                raise IndexError("row index out of range")
+        if j >= n:
+            raise IndexError("column index out of range")
+
+        try:
+            return self[i][j]
+        except KeyError:
+            return self.domain.zero
+
+    def extract_slice(self, slice1, slice2):
+        m, n = self.shape
+        ri = range(m)[slice1]
+        ci = range(n)[slice2]
+
+        sdm = {}
+        k = 0
+        for i, row in self.items():
+            if bool(row):
+                if i in ri:
+                    sdm[ri.index(i) + k] = {ci.index(j): e for j, e in row.items() if j in ci}
+            else:
+                k = k - 1
```

Comment:
What is k for?

Reply:
`k` helps to maintain the correct indices for the `sdm` dict.
For example, if we don't use `k`:
In the end, `sdm `will be `{2: {0: 1}}` instead of `{0: {0: 1}}`

Kindly share your opinion.
Thanks.

Reply:
I think this was better before. This implementation fails in several ways:
```python
In [12]: from sympy.polys.domainmatrix import DomainMatrix

In [13]: dM = DomainMatrix({2: {2: ZZ(1)}, 4: {4: ZZ(1)}}, (5, 5), ZZ)

In [14]: dM[:2,:2]
Out[14]: DomainMatrix([], (0, 0), ZZ)

In [15]: dM[2:,2:]
---------------------------------------------------------------------------
DDMBadInputError

In [16]: dM[3:,3:]
---------------------------------------------------------------------------
KeyError
```
I don't think that `k` is doing the right thing here but I can't test it because lots of examples seem to raise.

Reply:
> I think this was better before. This implementation fails in several ways:
> 
Can you please tell what should be the output of the following:
```
dM[:2,:2] : Should it be `DomainMatrix([], (0, 0), ZZ)` or `DomainMatrix([], (0, 2), ZZ)`?

dM[2:,2:]
```
Thanks.

Reply:
None of those!

You can test this with Matrix:
```python
In [11]: dM.to_Matrix()[:2,:2]
Out[11]: 
⎡0  0⎤
⎢    ⎥
⎣0  0⎦

In [12]: dM.to_Matrix()[2:,2:]
Out[12]: 
⎡1  0  0⎤
⎢       ⎥
⎢0  0  0⎥
⎢       ⎥
⎣0  0  1⎦
```

Reply:
Okay, now it's clear to me. I was removing those empty rows before instead of adding them in the DomainMatrix.
Thanks a lot.

> You can test this with Matrix:
>
Will keep it in mind. Thanks.